ASP.NET API/WAVC_WebApi/Hubs/FriendRequestHub.cs

```diff
@@ -0,0 +1,97 @@
+﻿using Microsoft.AspNetCore.Authorization;
+using Microsoft.AspNetCore.Identity;
+using Microsoft.AspNetCore.SignalR;
+using System;
+using System.Threading.Tasks;
+using WAVC_WebApi.Data;
+using WAVC_WebApi.Hubs.Interfaces;
+using WAVC_WebApi.Models;
+
+namespace WAVC_WebApi.Hubs
+{
+    [Authorize]
+    public class FriendRequestHub : Hub<IFriendRequestClient>
+    {
+        private readonly ApplicationDbContext _dbContext;
+        private readonly UserManager<ApplicationUser> _userManager;
+        private readonly FriendsManager _friendsManager;
+
+        public FriendRequestHub(UserManager<ApplicationUser> userManager, ApplicationDbContext dBContext)
+        {
+            _dbContext = dBContext;
+            _userManager = userManager;
+            _friendsManager = new FriendsManager(dBContext);
+        }
+
+        public async Task<bool> DeleteFriend(string id)
+        {
+            var sender = await _userManager.GetUserAsync(Context.User);
+            var reciever = _dbContext.Users.Find(id);
+
+            var result = await _friendsManager.DeleteFriendAsync(sender, reciever);
+            if (!result)
+            {
+                return false;
+            }
+            await Clients.User(reciever.Id).FriendDeleted(sender.Id);
+            return true;
+        }
+
+        public async Task SendFriendRequest(string id)
```

Comment:
I would move all these functions to a controller. There is a rule of thumb of good programming practices that says functions should have one responsibility.  In this example this function has at least two, moreover there is a mix of abstraction levels.

Reply:
To be honest, I'm not entirely sure how you want to do it, so I'd be glad if you made it yourself for at least one method.

Reply:
In FriendsRequestHub you should create a private property of the controller and invoke a desired function in here. At the bottom of the controller's function there should be an invocation of a method in a Hub to notify all clients. In that case no matter whether there was a change in a DB done by API call or signalR, all clients will be properly notified.